hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/records/RecordCache.java

```diff
@@ -37,79 +45,117 @@
-        final var receipts = getReceipts(transactionID);
-        return receipts.isEmpty() ? null : receipts.get(0);
+        /**
+         * This receipt is returned whenever we know there is a transaction pending (i.e. we have a history for a
+         * transaction ID), but we do not yet have a record for it.
+         */
+        private static final TransactionReceipt PENDING_RECEIPT =
+                TransactionReceipt.newBuilder().status(UNKNOWN).build();
+
+        /**
+         * Create a new {@link History} instance with empty lists.
+         */
+        public History() {
+            this(new HashSet<>(), new ArrayList<>(), new ArrayList<>());
+        }
+
+        /**
+         * Gets the primary record, that is, the record associated with the user transaction itself. This record
+         * will be associated with a transaction ID with a nonce of 0 and no parent consensus timestamp.
+         *
+         * @return The primary record, if there is one.
+         */
+        @Nullable
+        public TransactionRecord userTransactionRecord() {
+            return records.isEmpty() ? null : records.get(0);
+        }
+
+        /**
+         * Gets the primary receipt, that is, the receipt associated with the user transaction itself. This receipt will
+         * be null if there is no such record.
+         *
+         * @return The primary receipt, if there is one.
+         */
+        @Nullable
+        public TransactionReceipt userTransactionReceipt() {
+            return records.isEmpty() ? PENDING_RECEIPT : records.get(0).receipt();
+        }
+
+        /**
+         * Gets the list of all duplicate records, as a view. Should the list of records change, this view will reflect
+         * those changes.
+         *
+         * @return The list of all duplicate records.
+         */
+        @NonNull
+        public List<TransactionRecord> duplicateRecords() {
+            return records.isEmpty() ? emptyList() : records.subList(1, records.size());
+        }
+
+        /**
+         * Gets the number of duplicate records.
+         *
+         * @return The number of duplicate records.
+         */
+        public int duplicateCount() {
+            return records.isEmpty() ? 0 : records.size() - 1;
+        }
+
+        /**
+         * Returns a list of all records, ordered by consensus timestamp. Some elements of {@link #childRecords} may
+         * come before those in {@link #records}, while some may come after some elements in {@link #records}.
+         * @return The list of all records, ordered by consensus timestamp.
+         */
+        public List<TransactionRecord> orderedRecords() {
+            final var ordered = new ArrayList<>(records);
+            ordered.addAll(childRecords);
+            ordered.sort((a, b) -> TIMESTAMP_COMPARATOR.compare(
+                    a.consensusTimestampOrElse(Timestamp.DEFAULT), // Comparator doesn't currently handle nulls
```

Comment:
Given how easy it is to write a class implementing the `Comparator` interface (it's just one method), it might be worth doing that in a follow-up PR to handle nulls properly and remove the `somethingOrElse` calls that happen in dense loops.